lib/charms/smtp_integrator/v0/smtp.py

```
@@ -295,11 +295,20 @@ def _get_relation_data_from_relation(
         relation_data = relation.data[relation.app]
         if not relation_data:
             return None
+
+        password = relation_data.get("password")
+        if password is None and relation_data.get("password_id"):
+            password = (
+                self.model.get_secret(id=relation_data.get("password_id"))
+                .get_content()
+                .get("password")
```

What happens in case of an error? Should we catch an exception (e.g.  as done in https://github.com/canonical/github-runner-webhook-router/blob/main/charm/src/charm.py#L141-L144) here?

---

We talked about this with Christopher and we have 2 options:

1- Create a custom exception and raise it when you catch the `ops.model.ModelError`
2- Log the error and let the original error raise.

Having a custom exception would be nice and can be used by user to retry. Can be useful when infrastructure is unstable.
@jdkandersson wdyt?

---

Hmm, since this is a library I'm not sure how common it is to raise a custom exception? Do other libraries do that? Perhaps we check the data platforms and observability libraries to see. If they do, I'm fine with it, otherwise perhaps we just let it bubble up. In theory, everything should be working

---

Dataplatform has custom exceptions: https://github.com/canonical/postgresql-operator/blob/0dc4961bb2ae8c1bb26a1a5cadd222963479a33f/lib/charms/data_platform_libs/v0/data_interfaces.py#L378